src/qutip_cupy/expectation.py

```diff
@@ -0,0 +1,71 @@
+import cupy as cp
+
+
+def expect_cupydense(op, state):
+    """
+    Get the expectation value of the operator `op` over the state `state`.  The
+    state can be either a ket or a density matrix.
+
+    The expectation of a state is defined as the operation:
+        state.adjoint() @ op @ state
+    and of a density matrix:
+        tr(op @ state)
+    """
+    if state.shape[1] == 1:
+        return _expect_dense_ket(op, state)
+    return _expect_dense_dense_dm(op, state)
+
+
+def _expect_dense_ket(op, state):
+    _check_shape_ket(op, state)
+    return cp.vdot(state._cp, op._cp @ state._cp).item()
+
+
+def _check_shape_dm(op, state):
+    if (
+        op.shape[1] != state.shape[0]  # Matrix multiplication
+        or state.shape[0] != state.shape[1]  # State is square
+        or op.shape[0] != op.shape[1]  # Op is square
+    ):
+        raise ValueError(
+            "incorrect input shapes " + str(op.shape) + " and " + str(state.shape)
+        )
+
+
+_expect_dense_kernel = cp.RawKernel(
+    r"""
+    #include <cupy/complex.cuh>
+    extern "C" __global__
+    void expect_dens(const complex<double>* x1,const complex<double>* x2,
+                                        const int size, complex<double>* y) {
+        for (unsigned int tidx = blockDim.x * blockIdx.x + threadIdx.x; tidx < size;
+                                                      tidx += gridDim.x * blockDim.x) {
+            for(unsigned int j= 0; j<size; j++){
+                    y[tidx] += x1[j*size+tidx] * x2[tidx*size+j];
+                };
+        };
+    }""",
+    "expect_dens",
+)
+
+
+def _expect_dense_dense_dm(op, state):
+    _check_shape_dm(op, state)
+    size = op.shape[0]
+    out = cp.zeros((size,), dtype=cp.complex128)
+    # TODO: check if there is a better way to set thread dim and block dim
+    block_size = 64
```

Comment:
The execution speed is probably approximately linear in the number of blocks so we really should make the block size as high as possible. Since 2010, 1024 is allowed, so I suggest we pick that.

Reply:
The issue with 1024, is that it is the cap for many GPUs and you may end up with a problem if say for example you have only one GPU and you are running graphics there.

Reply:
Hmm. Perhaps let's just document that it a comment for now then?

Reply:
To be clear what actually happens is that using 1024 may take up too many registers and too much shared memory and not let a block from another process run in the same SM. But I'll leave it a t 1024 and leave the comment just in case.